### PR TITLE
Adding Buttplug.io Support

### PR DESCRIPTION
**Pull request by drone4770**
Adding support of Buttplug.io devices using buttplug ts library and connecting to an intiface-event server.

Server port has to be specified in .env with INTIFACE_PORT variable, otherwise, the module is not activated.

### CODE REVIEW

**Comment by drone4770** [comment]
What I don't understand for now is why the constructor of serialDeviceFactory gets called at this line: https://github.com/drone4770/slvctrlplus-server/blob/218e60e05e1dae28212e05890e0b145041fe632a/src/device/buttplugIoDeviceProvider.ts#L54

File: src/device/buttplugIoDeviceProvider.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    public constructor(eventEmitter: EventEmitter, deviceFactory: ButtplugIoDeviceFactory) {
+        super(eventEmitter);
+        this.buttplugIoDeviceFactory = deviceFactory;
+    }
+
+    public init(): void
+    {
+        if (process.env.INTIFACE_PORT) {
+            this.buttplugConnector = new ButtplugNodeWebsocketClientConnector("ws://127.0.0.1:" + process.env.INTIFACE_PORT + "/buttplug");
```
Comment (heavyrubberslave) [inline src/device/buttplugIoDeviceProvider.ts:28]:
```suggestion
            this.buttplugConnector = new ButtplugNodeWebsocketClientConnector(`ws://127.0.0.1:${process.env.INTIFACE_PORT}/buttplug`);
```

File: src/device/buttplugIoDeviceProvider.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    private readonly buttplugIoDeviceFactory: ButtplugIoDeviceFactory;
+
+    public constructor(eventEmitter: EventEmitter, deviceFactory: ButtplugIoDeviceFactory) {
+        super(eventEmitter);
+        this.buttplugIoDeviceFactory = deviceFactory;
+    }
+
+    public init(): void
+    {
+        if (process.env.INTIFACE_PORT) {
```
Comment (heavyrubberslave) [inline src/device/buttplugIoDeviceProvider.ts:27]: I think eventually we should configure the different providers in `settings.json` and only instantiate the provider if it's actually configured there. Will adjust my POC implementation in the coming days. Eventually there should be a user interface under "Settings" menu point to add/edit/remove providers.

Comment (drone4770) [reply src/device/buttplugIoDeviceProvider.ts:27]: I didn't know how to do that at first this is why I used the port in env as a way to control that, but I was not really happy with this either

Comment (heavyrubberslave) [reply src/device/buttplugIoDeviceProvider.ts:28]: My energy level is currently rather low so I don’t spend too much time working on SlvCtrl+ currently. But today I started with the refactoring. I did good progress but it needs another few hours of work until it’s done. Sorry for the delay.

**Comment by heavyrubberslave** [comment]
Hello @drone4770, not sure if you're still on this but I finally found time to implement a proper approach for configurable device sources (haven't updated the UI yet though, but you can configure them by editing the settings.json in your home directory).

See: https://github.com/SlvCtrlPlus/slvctrlplus-server/pull/13

You need to delete the settings file as it's not backward compatible.

It looks now like this:

```json
{
    "knownDevices": {
        "35966B6A50543539322E3120FF110A0A": {
            "id": "3436db9f-6fd7-4a00-ad71-7a316b265d83",
            "serialNo": "35966B6A50543539322E3120FF110A0A",
            "name": "Roasted Utapau",
            "type": "dimmer",
            "source": "slvCtrlPlus",
            "config": {}
        },
        "9CC9E44350543539322E3120FF0F093B": {
            "id": "ade55458-7143-4bfe-955a-5b8a45441495",
            "serialNo": "9CC9E44350543539322E3120FF0F093B",
            "name": "Empty Alderaan",
            "type": "et312",
            "source": "slvCtrlPlus",
            "config": {}
        }
    },
    "deviceSources": {
        "b6a0f45e-c3d0-4dca-ab81-7daac0764291": {
            "id": "b6a0f45e-c3d0-4dca-ab81-7daac0764291",
            "type": "slvCtrlPlusSerial",
            "config": {}
        }
    }
}
```

where `config` is just a JSON object where you can put source specific configs (for example the URL for the buttplug.io server)

**Comment by heavyrubberslave** [comment]
@drone4770 I merged the configurable device sources to main, that's why the PR was closed. Feel free to open a new one against main after rebasing your feature branch.

**Pull request closed without merging**